main.tf

```
@@ -38,6 +38,27 @@ resource "google_storage_bucket" "gcs_bucket" {
       cors             # ignore CORS changes. Use 'gsutil' tool instead. See https://cloud.google.com/storage/docs/configuring-cors
     ]
   }
+  dynamic "lifecycle_rule" {
+    for_each = var.lifecycle_rules
+    content {
+      action {
+        type          = lifecycle_rule.value.action.type
+        storage_class = lookup(lifecycle_rule.value.action, "storage_class", null)
+      }
+      condition {
+        age            = lookup(lifecycle_rule.value.condition, "age", null)
+        created_before = lookup(lifecycle_rule.value.condition, "created_before", null)
+        with_state = lookup(
+          lifecycle_rule.value.condition, "with_state",
+          lookup(lifecycle_rule.value.condition, "is_live", false) ? "LIVE" : null
+        )
+        matches_storage_class = contains(keys(lifecycle_rule.value.condition), "matches_storage_class") ? split(
+          ",", lifecycle_rule.value.condition["matches_storage_class"]
+        ) : null
```

we can just do `matches_storage_class = lookup(lifecycle_rule.value.condition, "matches_storage_class", null)`

and provide the value of `lifecycle_rule.value.condition["matches_storage_class"]` as a list of strings instead of single a comma-delimited string..

---

conditions type is map(string) can not have mixed, that's why it is comma delimited string

---

Ok. Makes sense. If that's the case, then in true case it will create an array of strings and in false case it should create an empty array instead of null. Keeps the data structure coherent.

---

an empty array still value, it should be null.

---

```suggestion
        matches_storage_class = contains(keys(lifecycle_rule.value.condition), "matches_storage_class") ? (
          split(",", lifecycle_rule.value.condition["matches_storage_class"])
        ) : null
```
to read the full `split()` command in a single line